### ISSUE TITLE
long ligatures get shorter

### ISSUE THREAD

**Issue by OldWorldOrdr** [kovidgoyal/kitty#6220]
[test.txt](https://github.com/kovidgoyal/kitty/files/11349487/test.txt)
<img width="1026" alt="Screen Shot 2023-04-27 at 11 07 17 PM" src="https://user-images.githubusercontent.com/57970114/235045583-385c3043-7b7f-466e-922a-0bcc8fbb2223.png">
the a is right after the - and the = lines, there shouldn't be a gap there

I've tested with windows terminal and it renders correctly, so does termux

**Comment by kovidgoyal** [comment]
That will be because whatever font is being used for the ligature is not producing monospaced ligatures, aka the ligature glyph's width is not the same as the width of n*char_width. Report the issue to the font developer.

**Comment by OldWorldOrdr** [comment]
If that's the case why do windows terminal and termux render correctly with the same font?

**Comment by kovidgoyal** [comment]
No clue, maybe they special case rendering or dont actually use the font
or some other hack along those lines.

**Comment by Finii** [comment]
The lig's glyphs are all the same advance width.
The negative bearings for the middle `=` is even bigger than for the end `=`.

![image](https://user-images.githubusercontent.com/16012374/235410622-47d5c0a9-d6c5-4e2a-b37d-6ffcbe67cbe0.png)

**Comment by kovidgoyal** [comment]
And yet the N glyphs when positioned by harfbuzz end up being shorter
than the N individual glyphs. I didnt look at it in detail but most likely
it will be because the individual glyphs are not full width, that is,
they dont occupy the same width as the widest character in the font.
This doesnt matter when rendering individual glyphs since kitty
compensates for this automatically, placing the glyph in a fixed width
cell. But when N such glyphs are rendered, the end result will be
significantly less than N * monospace width. So let me amend my previous
sentence slightly, the ligature glyphs need to be as wide as the widest
char if they are going to have any hope of lining up.

And on a more general note, as I have said before, this is entirely the
wrong approach for getting continuous lines that line up properly in
terminals. The correct solution is to use Unicode box drawing
characters, they are font independent and always line up.

**Comment by Finii** [comment]
Thanks for the response, it clears that a bit up! :+1:

> Unicode box drawing characters

Not really a solution if you want to get

![image](https://user-images.githubusercontent.com/16012374/235866499-0fbd1f10-eddc-4683-bcf0-6aa50838f927.png)
_Screenshot: Scribus_

I guess. Can not test, because my `kitty` does not render ligs at all. Need to pull HEAD and rebuild :-)

But then the question remains if such 'tricks' are needed in a terminal... :thinking: I personally never used these features.

**Comment by Finii** [comment]
Btw, it renders 100% ok in my `kitty`:

![image](https://user-images.githubusercontent.com/16012374/235891439-dbd905d3-e5f7-4483-9e65-94c6a97e063a.png)

So this is neither a font bug nor a kitty bug ;-D

P.S. Built my `kitty` from
```
8f96395f7 (HEAD -> master, origin/master, origin/HEAD) diff kitten: Fix a regression in 0.28.0 that broke using relative paths as arguments to the kitten
```

**Comment by kovidgoyal** [comment]
Cool, am glad it works. Thanks for testing. long arrows dont need to line up exactly, as it doesnt matter if there is abit extra space between the arrow head and whatever char comes next, so they work fine anyway.

Lineups are required most commonly for table rendering. For that usecase unicode box drawing is definitely the way to go.
